crates/dbsp/src/trace/spine_async/thread.rs

```diff
@@ -0,0 +1,127 @@
+//! A compactor thread that merges the batches for the spine-fueled trace.
+
+use crate::Runtime;
+use std::cell::RefCell;
+use std::mem::{replace, take};
+use std::sync::{Arc, Mutex, Weak};
+use std::thread::{Builder, Thread};
+use std::thread_local;
+
+/// Return value for a worker function.
+pub enum WorkerStatus {
+    /// The worker has more work to do (it only returned to allow other workers
+    /// to run).
+    Busy,
+
+    /// The worker has no more work to do now, but it might have more later.
+    Idle,
+
+    /// The worker has exited.
+    Done,
+}
+
+struct Inner {
+    new_workers: Vec<Box<dyn FnMut() -> WorkerStatus + Send>>,
+    exiting: bool,
+    thread: Option<Thread>,
+}
+pub struct BackgroundThread(Mutex<Inner>);
+
+// There are three cases:
+//
+// 1. We are a worker thread inside a `Runtime`. The background thread
+//    should correspond to the worker thread. We accomplish this.
+//
+// 2. We are part of a circuit that is not part of a `Runtime`. We would
+//    prefer to have a background thread for the circuit. We do not
+//    accomplish this; instead, we have a background thread for the
+//    thread in which the circuit exists (one could have more than one
+//    circuit per thread). However, since a `ChildCircuit` is not
+//    `Send`, at least we will have only one background thread per
+//    circuit.
+//
+// 3. We are not part of a circuit at all; that is, something
+//    instantiated a `Spine` outside a circuit (probably in a unit
+//    test). We create a background thread for the thread that initially
+//    owned the `Spine`.  (`Spine` is not `Send` either.)
+//
+// It doesn't make sense to tie the background thread to the `Runtime`
```

Comment:
I guess the long term question is if we want to go towards a world where we always have a runtime or live in this (mad) world where we can have three case. I'd much favor always having a runtime as it really complicates the code for unnecessary reasons. (it's not something to we can fix in this PR though)

Reply:
I hate not always having a runtime.

Reply:
Do we not always have the runtime because we have an API to create a circuit without one?  Or is it because, the thread running this code can be outside the runtime?

Reply:
Both reasons:

1. We can have a spine outside a runtime because a circuit can be created without a runtime.
2. We can have a spine outside a runtime because we instantiated the spine without a circuit at all (for testing).

If we didn't have case 1, then I'd just get rid of case 2 (by rewriting the spine tests to run inside a circuit), but case 1 means there's no value in that.

Reply:
I think it's ok to get rid of case 1.

Reply:
I opened https://github.com/feldera/feldera/issues/2117. It's too much for this PR.